### PR TITLE
ddl: Only warn of subpartitions, still allow first level partitioning

### LINKED ISSUE #41198: List Partition table doesn't support subpartition but is created successfully without any warnings
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
DROP TABLE IF EXISTS thash14;
CREATE TABLE thash14 (
    col1 INT NOT NULL,
    col2 INT NOT NULL,
    col3 INT NOT NULL,
    col4 INT NOT NULL,
    primary KEY (col1,col3)
)
PARTITION BY LIST(col1)
SUBPARTITION BY HASH(col3) SUBPARTITIONS 2 
(
	PARTITION p1 VALUES IN (1, 2, 3, 4, 5),
	PARTITION p2 VALUES IN (6, 7, 8, 9, 10),
	PARTITION p3 VALUES IN (11, 12, 13, 14, 15)
);
```

### 2. What did you expect to see? (Required)
table is not created, or created with a warning that indicates list partition table doesn't support subpartition.

### 3. What did you see instead (Required)
table created successfully.
### 4. What is your TiDB version? (Required)

Release Version: v6.2.0-alpha-2066-g62e8709-dirty
Edition: Community
Git Commit Hash: 62e87090e9d008aeb8e920347fba541c15f7aa9f
Git Branch: key-partition
UTC Build Time: 2023-02-08 07:26:38
GoVersion: go1.19
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: tikv

### LINKED ISSUE #41200: Partition table of TiDB doesn't support subpartition but is created successfully
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
drop table if exists thash15;
CREATE TABLE thash15 (
    col1 INT NOT NULL,
    col2 INT NOT NULL,
    col3 INT NOT NULL,
    col4 INT NOT NULL,
    primary KEY (col1,col3)
)
PARTITION BY HASH(col1) PARTITIONS 4
SUBPARTITION BY HASH(col3) SUBPARTITIONS 2 
```

### 2. What did you expect to see? (Required)
The table is not created and give an error that is "It is only possible to mix RANGE/LIST partitioning with HASH/KEY partitioning for subpartitioning"
### 3. What did you see instead (Required)
The table is created bu treated as normal table.
### 4. What is your TiDB version? (Required)

 Release Version: v6.2.0-alpha-2066-g62e8709-dirty
Edition: Community
Git Commit Hash: 62e87090e9d008aeb8e920347fba541c15f7aa9f
Git Branch: key-partition
UTC Build Time: 2023-02-08 07:26:38
GoVersion: go1.19
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: tikv

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: close #41198 , close #41200

Problem Summary:
For LIST partitioning, it did not check if SUBPARTITION was used and just stripped the subpartition part
For RANGE and HASH partitioning it stripped all partitioning and added a warning that such partitioning was not supported.

### What is changed and how it works?
Combine the two behaviours:
Strip the subpartitioning from the CREATE TABLE statement and issue a warning.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
If CREATE TABLE includes subpartitions, keep the normal partitioning and skip the subpartitioning + add a warning.
```

### CODE REVIEW

**Comment by mjonss** [comment]
Also fixes #42000
```
tidb> CREATE TABLE thash15 ( col1 INT NOT NULL, col2 INT NOT NULL, col3 INT NOT NULL, col4 INT NOT NULL, primary KEY (col1,col3) ) PARTITION BY HASH(col1) PARTITIONS 4 SUBPARTITION BY HASH(col3) SUBPARTITIONS 2;
Query OK, 0 rows affected, 1 warning (0.04 sec)

tidb> show warnings;
+---------+------+-----------------------------------------------------------+
| Level   | Code | Message                                                   |
+---------+------+-----------------------------------------------------------+
| Warning | 8200 | Unsupported subpartitioning, only using HASH partitioning |
+---------+------+-----------------------------------------------------------+
1 row in set (0.00 sec)
```

**Comment by TonsnakeLin** [comment]
I think it is reasonable for #41198, but it's not reasonable for #41200 which gives an error like "It is only possible to mix RANGE/LIST partitioning with HASH/KEY partitioning for subpartitioning" in MySQL.

**Comment by TonsnakeLin** [comment]
In addition, should we modify the document about partition if we fixed the issue.

**Comment by mjonss** [comment]
> I think it is reasonable for #41198, but it's not reasonable for #41200 which gives an error like "It is only possible to mix RANGE/LIST partitioning with HASH/KEY partitioning for subpartitioning" in MySQL.

I have changed the logic so it will give that error if subpartitioning a HASH or KEY partitioned table, just like MySQL.

File: ddl/partition.go [hunk 1]
```diff
@@ -440,35 +440,32 @@ func buildTablePartitionInfo(ctx sessionctx.Context, s *ast.PartitionOptions, tb
...
-		// Partition by list is enabled only when tidb_enable_list_partition is 'ON'.
-		enable = ctx.GetSessionVars().EnableListTablePartition
 	}
 
 	if !enable {
 		ctx.GetSessionVars().StmtCtx.AppendWarning(dbterror.ErrUnsupportedCreatePartition.GenWithStack(fmt.Sprintf("Unsupported partition type %v, treat as normal table", s.Tp)))
 		return nil
 	}
+	if s.Sub != nil {
+		ctx.GetSessionVars().StmtCtx.AppendWarning(dbterror.ErrUnsupportedCreatePartition.GenWithStack(fmt.Sprintf("Unsupported subpartitioning, only using %v partitioning", s.Tp)))
```
Comment (TonsnakeLin) [inline ddl/partition.go:467]: remove this empty line

Comment (TonsnakeLin) [reply ddl/partition.go:467]: Add space indentation at the beginning of the line

**Review by TonsnakeLin — APPROVED** [review]
LGTM execep the empty line and spaces I commented

**Review by Defined2014 — APPROVED** [review]

**Pull request merged**